Add HexaryTrie.traverse() and .traverse_from()

carver opened the pull request:
These are useful for inspecting trie structure (like when walking the
database to validate presence of all nodes). Traverse_from in particular
is helpful for caching, so you don't need to re-navigate from the root
every time.

TODO:
- [x] test `traverse_from` `MissingTraversalNode`
- [x] drop allowing requested_key = None
- [x] fix changelog for ^
- [x] new changelog entry
- [x] `missing_node_hash` of `MissingTrieNode` to `HexBytes` also
- [x] trie.root_node -> `MissingTraversalNode` error
- [x] test ^
- [x] CAP
- [x] issue to annotate root_node #103 
- [x] squash lots of commits

#### Cute Animal Picture

![Cute animal picture](https://static.boredpanda.com/blog/wp-content/uploads/2020/04/cat-and-dog-challenge-fb36-png__700.jpg) *It's possible I'm the dog*

carver committed:
Add HexaryTrie.traverse() and traverse_from

These are useful for inspecting trie structure (like when walking the
database to validate presence of all nodes). Traverse_from in particular
is helpful for caching, so you don't need to re-navigate from the root
every time.

- Returns new annotated HexaryTrieNode from public traversal API
- Raise new MissingTraversalNode, analogous to MissingTrieNode
- Raise new TraversedPartialPath from .traverse() and traverse_from()

carver commented on trie/hexary.py:
```diff
@@ -144,6 +162,27 @@ def _traverse(self, root_hash, trie_key) -> Tuple['Node', Tuple[int, ...]]:
+
+        The trie does *not* re-verify the path/hashes from the node prefix to the node.
+        """
+        trie_key = Nibbles(trie_key_input)
```
note to self: do the same thing for `node_prefix`

carver commented on trie/hexary.py:
```diff
@@ -144,6 +162,27 @@ def _traverse(self, root_hash, trie_key) -> Tuple['Node', Tuple[int, ...]]:
+        # For now, traverse will raise if you try to traverse into
+        # part of an extension node, or leaf node.
+        if len(remaining_key) > 0:
+            raise Exception(f"Tried to traverse {remaining_key!r} past {node!r}")
```
TODO: Maybe just return with a remaining key?

I think raising an exception is insufficient. Example: you have explored a trie from one root, knowing that you only need to explore the unmodified parts of the trie. Then the root changes and you want to continue exploring the unexplored parts. But due to possible changes, you might end up in the middle of an extension node on the next traversal.

carver replied:
With `traverse_from` it shouldn't be an issue, since you're always navigating from a fixed root (by supplying the parent node).

With `traverse()` though, it's definitely a possibility, and I still need to think through how to handle all the cases properly.

Though... maybe TODO: add root hash to traverse_from() args, to fix the `MissingTrieNode` exception... or make the root hash optional?

pipermerriam replied:
I'm very likely missing something but did you give any thought to merely returning a representation of the extension node for all paths that don't get you deep enough to hit one of the child nodes?

carver replied:
So far, the latest code has an API (direction) that I'm happy with. Traversal is all about knowing where nodes are in the trie, and moving to them. So if you go somewhere that isn't a node (like the middle of an extension or leaf), then it should raise an exception, complete with the information you need to find out what you need to keep traversing (the parent extension node, and the path to that).

carver commented on CHANGELOG:
```diff
@@ -12,8 +12,6 @@ Breaking Changes
   of returning b'' when a key is not present in the trie
   https://github.com/ethereum/py-trie/pull/98
-- If traversing, instead of getting a key value, then MissingTrieNode.requested_key might be None
-  https://github.com/ethereum/py-trie/pull/98
```
I decided this was too messy, and that a different exception with known values makes more sense (`MissingTraversalNode`)

pipermerriam commented on trie/hexary.py:
```diff
@@ -131,20 +153,88 @@ def _get(self, root_hash, trie_key):
+        Find the node at the path of nibbles provided. Unspecified behavior if
+        path ends in the middle of an extension.
+
+        :raises MissingTraversalNode: if a node body is missing from the database
```
As well as if the path ends in an extension node?

carver replied:
That's the `TraversedPartialPath` one line below

pipermerriam commented on trie/hexary.py:
```diff
@@ -131,20 +153,88 @@ def _get(self, root_hash, trie_key):
+        else:
+            return self._annotate_node(node)
+
+    @classmethod
```
this could be a static method or even a standalone util function?  I don't see any access to `cls`

pipermerriam commented on trie/typing.py:
```diff
@@ -33,3 +39,32 @@ def __new__(cls, nibbles):
             return tuple.__new__(cls, (Nibble(maybe_nibble) for maybe_nibble in nibbles))
+
+
+class TrieNode(NamedTuple):
```
Should this be `HexaryTrieNode`?

pipermerriam commented on trie/hexary.py:
```diff
@@ -131,20 +154,58 @@ def _get(self, root_hash, trie_key):
         return self._traverse_from(root_node, trie_key)
 
-    def _traverse_from(self, node, trie_key) -> Tuple['Node', Tuple[int, ...]]:
+    def traverse_from(self, node, trie_key_input: Nibbles) -> HexaryTrieNode:
```
type annotation for `node`?

pipermerriam commented on trie/hexary.py:
```diff
@@ -131,20 +154,58 @@ def _get(self, root_hash, trie_key):
             raise Exception("Invariant: This shouldn't ever happen")
 
-    def _traverse(self, root_hash, trie_key) -> Tuple['Node', Tuple[int, ...]]:
+    def traverse(self, trie_key_input: Nibbles) -> HexaryTrieNode:
```
Not clear to me what `trie_key_input` is.  Also it is typed as `Nibbles` but then the first thing this function does is cast this value to `Nibbles(trie_key_input)`.  I'm not understanding what or why that is happening.

carver replied:
`Nibbles` is just a nice way to validate that each int in the tuple is in `range(16)`, if someone passes in just a tuple of ints (since type hints are just a hint, not formally required). I'll add a comment about it being for validation.

pipermerriam commented on trie/utils/nodes.py:
```diff
@@ -173,3 +176,43 @@ def encode_leaf_node(value):
     return LEAF_TYPE_PREFIX + value
+
+
+def annotate_node(node_body):
```
type hints since this is new code?

carver replied:
Okay, yeah, I'll look at how carefully I can define the node bodies. I think it's a recursive type definition, which I don't know/remember how to do.

pipermerriam reviewed (approved):
Looks like `README` needs to be updated to document the new traversal APIs.

As for the actual traversal mechanism, I can't claim that I fully ingested it, but structurally/architecturally things look sound.

carver committed:
Write up some traversal docs

Also, the suffix and sub_segments were not being cast to Nibbles! Fixed
that bug, which became apparent during the python repr printouts.

carver commented on README.md:
```diff
@@ -108,6 +106,44 @@ True
+# Look at the root node:
+>>> root_node = t.traverse(())
+>>> root_node
+HexaryTrieNode(sub_segments=((0x6, 0xd, 0x7, 0x9, 0x2, 0xd, 0x6),), value=b'', suffix=(), raw=[b'\x16\xd7\x92\xd6', b'\xb4q\xb8h\xec\x1c\xe1\xf4\\\x88\xda\xb4\xc1\xc2n\xbaw\xd0\x9c\xf1\xacV\xb4Dk\xa7\xe6\xd7qf\xc2\x82'])
```
Thanks for recommending the docs. Here it is. Hopefully it gives more of a feel for how the traversal API works in practice.

(Not specifically seeking feedback, just thought you might be interested)
